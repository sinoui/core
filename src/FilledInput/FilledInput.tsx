import React, { useState, useCallback } from 'react';
import BaseInput, { BaseInputProps } from '@sinoui/core/BaseInput';
import classNames from 'classnames';
import styled, { css } from 'styled-components';

export interface FilledInputProps extends BaseInputProps {
  error?: string;
  warning?: boolean;
  focused?: boolean;
}

const FULL_SCALE_X = 1;
const FULL_SCALE_Y = 1;

const transformStyle = (props: { focused?: boolean }) => {
  const scaleX = props.focused ? FULL_SCALE_X : 0;
  const scaleY = props.focused ? FULL_SCALE_Y : 0;
  return css`
    transform: scaleX(${scaleX}) scaleY(${scaleY});
  `;
};

const errorOrWarnigStyle = css<FilledInputProps>`
  && {
    &::after {
      background-color: ${(props) =>
        props.theme.palette[props.error ? 'error' : 'warning'][500]};
      transform: scaleX(${FULL_SCALE_X}) scaleY(${FULL_SCALE_Y});
    }
  }
`;

const disabledunderlineStyle = css`
  &::after {
    height: 0px;
  }

  &::before {
    height: 0px;
  }
`;

const inkbarStyle = css`
  &::after {
    background-color: ${(props) =>
      props.theme.palette.primary[
        props.theme.palette.type === 'light' ? 700 : 300
      ]};
    left: 0;
    bottom: 0;
    content: '';
    height: 2px;
    position: absolute;
    right: 0;
    transition: ${(props) =>
      props.theme.transitions.create('transform', {
        duration: props.theme.transitions.duration.shorter,
        easing: props.theme.transitions.easing.easeOut,
      })};
    pointer-events: none;
    ${transformStyle};
  }
`;

const underlineStyle = css<FilledInputProps>`
  &::before {
    background-color: ${(props) => props.theme.palette.input.bottomLine};
    left: 0;
    bottom: 0;
    content: '';
    display: block;
    height: 1px;
    position: absolute;
    right: 0;
    transition: ${(props) =>
      props.theme.transitions.create('background-color', {
        duration: props.theme.transitions.duration.shorter,
        easing: props.theme.transitions.easing.easeInOut,
      })};
    pointer-events: none;
    transform: scaleX(${FULL_SCALE_X}) scaleY(${FULL_SCALE_Y});
  }

  &:hover::before {
    ${(props) =>
      !props.disabled &&
      !props.readOnly &&
      !props.error &&
      `background-color: ${props.theme.palette.text.primary};
  `};
  }
`;

const disabledBackgroundStyle = css`
  background-color: ${(props) => props.theme.palette.text.divider};
`;

const StyledBaseInput = styled(BaseInput)<
  BaseInputProps & { error?: boolean; warning?: boolean }
>`
  position: relative;
  background-color: ${(props) =>
    props.theme.palette.type === 'light'
      ? 'rgba(0,0,0,0.09)'
      : 'rgba(255,255,255,0.09)'};
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  transition: ${(props) =>
    props.theme.transitions.create('background-color', {
      duration: props.theme.transitions.duration.shorter,
      easing: props.theme.transitions.easing.easeOut,
    })};

  &:hover {
    ${(props) =>
      !props.disabled &&
      !props.readOnly &&
      !props.error &&
      `background-color: ${
        props.theme.palette.type === 'light'
          ? 'rgba(0,0,0,0.13)'
          : 'rgba(255,255,255,0.13)'
      };
      cursor:pointer;
  `};
  }
  ${(props) => !props.disabled && inkbarStyle};
  ${(props) => !props.error && underlineStyle};
  ${(props) => (props.error || props.warning) && errorOrWarnigStyle};
  ${(props) => props.disabled && disabledunderlineStyle};
  ${(props) => props.disabled && disabledBackgroundStyle};

  > input {
    padding: 27px 12px 10px;
  }
`;

export default function FilledInput(props: FilledInputProps) {
  const {
    fullWidth = false,
    inputComponent = 'input',
    multiline = false,
    type = 'text',
    ref,
    className,
    error,
    warning,
    ...other
  } = props;
  const [focused, setFocused] = useState(false);

  const handleBlur = useCallback(
    (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      if (!props.readOnly) {
        setFocused(false);
        if (props.onBlur) {
          props.onBlur(event);
        }

        if (props.inputProps && props.inputProps.onBlur) {
          props.inputProps.onBlur(event);
        }
      }
    },
    [props],
  );

  const handleFocus = useCallback(
    (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      if (!props.readOnly) {
        setFocused(true);
        if (props.onFocus) {
          props.onFocus(event);
        }

        if (props.inputProps && props.inputProps.onFocus) {
          props.inputProps.onFocus(event);
        }
      }
    },
    [props],
  );

  return (
    <StyledBaseInput
      fullWidth={fullWidth}
      error={!!error}
      warning={warning}
      focused={focused}
      inputComponent={inputComponent}
      multiline={multiline}
      type={type}
      ref={ref}
      {...other}
      className={classNames('sinoui-filled-input', className)}
      onFocus={handleFocus}
      onBlur={handleBlur}
    />
  );
}
