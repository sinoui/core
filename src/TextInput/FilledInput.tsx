import type { BaseInputProps } from '@sinoui/core/BaseInput';
import BaseInput from '@sinoui/core/BaseInput';
import React from 'react';
import styled, { css } from 'styled-components';

import { FILLED_INPUT_BGCOLOR } from './constant';
import lineRippleStyle from './lineRippleStyle';

/**
 * 填充模式下输入框组件属性类型
 */
export interface FilledInputProps extends BaseInputProps {
  /**
   * 聚焦状态
   */
  focused?: boolean;
  /**
   * 密集模式
   */
  dense?: boolean;
  /**
   * 无标签
   */
  noLabel?: boolean;
}

const denseStyle = css`
  padding-top: 20.5px;
  padding-bottom: 3.5px;
`;

const denseNoLabelStyle = css`
  padding-top: 3.5px;
  padding-bottom: 3.5px;
`;

const overlayStyle = css<{
  /**
   * 组件处于聚焦状态
   */
  focused?: boolean;
}>`
  > .sinoui-input__overlay {
    position: absolute;
    width: 100%;
    height: 100%;
    top: 0%;
    left: 0%;
    background-color: ${({ theme }) => theme.palette.text.primary};
    pointer-events: none;
    z-index: 2;
    transition: ${({ theme: { transitions } }) =>
      transitions.create('background-color', {
        easing: transitions.easing.easeInOut,
        duration: transitions.duration.shortest,
      })};
    opacity: ${({ focused }) => (focused ? 0.12 : 0)};
  }
  ${({ focused }) =>
    !focused &&
    `
      :hover > .sinoui-input__overlay {
        opacity: 0.04;
      }
  `}
`;

const FilledInputLayout = styled(BaseInput)<FilledInputProps>`
  position: relative;
  background-color: ${({ theme }) => FILLED_INPUT_BGCOLOR[theme.palette.type]};
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  transition: ${({ theme: { transitions } }) =>
    transitions.create('background-color', {
      duration: transitions.duration.shorter,
      easing: transitions.easing.easeOut,
    })};

  > .sinoui-input-adornment--end {
    margin-right: ${({ multiline }) => (multiline ? 0 : 12)}px;
    margin-left: -6px;
  }

  > .sinoui-input-adornment--start {
    margin-right: 0px;
    margin-left: ${({ multiline }) => (multiline ? -4 : 10)}px;
  }

  > .sinoui-input-adornment--start ~ .sinoui-base-input__input {
    margin-left: -6px;
  }

  > .sinoui-input-adornment--start.sinoui-input-adornment--text
    ~ .sinoui-base-input__input,
  > .sinoui-input-adornment--end.sinoui-input-adornment--text {
    margin-left: ${({ multiline }) => (multiline ? 0 : -12)}px;
  }

  > .sinoui-input-adornment--start,
  > .sinoui-input-adornment--text {
    transform: translateY(${({ noLabel }) => (noLabel ? 0 : 7.5)}px);
  }

  ${({ multiline }) => (multiline ? '&' : '> .sinoui-base-input__input')} {
    padding: 22px 12px 10px;
    padding-top: ${({ noLabel }) => (noLabel ? 7.5 : 25.5)}px;
    ${({ noLabel, dense }) => !noLabel && dense && denseStyle}
    ${({ noLabel, dense }) => dense && noLabel && denseNoLabelStyle}
  }

  ${lineRippleStyle}
  ${({ disabled }) => !disabled && overlayStyle}
`;

/**
 * 填充模式输入框
 */
const FilledInput = React.forwardRef<HTMLDivElement, FilledInputProps>(
  function FilledInput({ children, ...props }, ref) {
    return (
      <FilledInputLayout {...props} ref={ref}>
        {children}
        <div className="sinoui-input__overlay" />
      </FilledInputLayout>
    );
  },
);

export default FilledInput;
