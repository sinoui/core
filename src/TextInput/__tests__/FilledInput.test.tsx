import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import FilledInput from '../FilledInput';

it('正确渲染填充模式的input', () => {
  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <FilledInput />
    </ThemeProvider>,
  ).toJSON;

  expect(tree).toMatchSnapshot();
});

it('渲染不可用状态下的填充模式的Input', () => {
  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <FilledInput disabled />
    </ThemeProvider>,
  ).toJSON;

  expect(tree).toMatchSnapshot();
});

it('渲染错误状态下填充模式的Input', () => {
  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <FilledInput error />
    </ThemeProvider>,
  ).toJSON;

  expect(tree).toMatchSnapshot();
});

it('聚焦状态', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <FilledInput focused />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('无标签', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <FilledInput noLabel />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('密集模式', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <>
          <FilledInput dense />
          <FilledInput dense noLabel />
        </>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
