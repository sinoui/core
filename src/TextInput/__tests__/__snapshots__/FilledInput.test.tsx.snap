// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`正确渲染填充模式的input 1`] = `[Function]`;

exports[`渲染不可用状态下的填充模式的Input 1`] = `[Function]`;

exports[`渲染密集状态下填充模式的Input 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: text;
  color: rgba(0,0,0,0.87);
  font-size: 1rem;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  line-height: 1.1875em;
}

.c0 > input,
.c0 > textarea {
  padding: 6px 0 7px;
  display: block;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  line-height: inherit;
  font: inherit;
  color: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  border: none;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  -webkit-tap-highlight-color: transparent;
  min-width: 0px;
  width: 100%;
  height: 1.1876em;
  min-height: 1.1876em;
}

.c0 > input::-moz-placeholder,
.c0 > textarea::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-webkit-input-placeholder,
.c0 > textarea::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input:-ms-input-placeholder,
.c0 > textarea:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-ms-clear,
.c0 > textarea::-ms-clear {
  display: none;
}

.c0 > input:focus,
.c0 > textarea:focus {
  outline: 0;
}

.c0 > input:invalid,
.c0 > textarea:invalid {
  box-shadow: none;
}

.c0 > textarea {
  resize: none;
  overflow: hidden;
}

.c1 {
  position: relative;
  background-color: #f5f5f5;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  -webkit-transition: background-color 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  transition: background-color 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
}

.c1:hover {
  background-color: rgba(0,0,0,0.08);
  cursor: pointer;
}

.c1::after {
  background-color: #3f51b5;
  left: 0;
  bottom: 0;
  content: '';
  height: 2px;
  position: absolute;
  right: 0;
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  -webkit-transition: transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  transition: transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  pointer-events: none;
  -webkit-transform: scaleX(0) scaleY(0);
  -ms-transform: scaleX(0) scaleY(0);
  transform: scaleX(0) scaleY(0);
}

.c1::before {
  background-color: rgba(0,0,0,0.42);
  left: 0;
  bottom: 0;
  content: '';
  display: block;
  height: 1px;
  position: absolute;
  right: 0;
  -webkit-transition: background-color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: background-color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  pointer-events: none;
  -webkit-transform: scaleX(1) scaleY(1);
  -ms-transform: scaleX(1) scaleY(1);
  transform: scaleX(1) scaleY(1);
}

.c1:hover::before {
  background-color: rgba(0,0,0,0.87);
}

.c1 > input,
.c1 > textarea {
  padding: 27px 12px 10px;
  padding-top: 23px;
  padding-bottom: 6px;
}

<div
  className="c0 sinoui-base-input__layout c1 sinoui-filled-input"
  data-testid="baseInput"
  onClick={[Function]}
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`渲染错误状态下填充模式的Input 1`] = `[Function]`;
